PROTON-1408 -- fix for msg id rollover bug

mgoulish opened the pull request:
I'm getting a different hang now, but it looks like this part of the code is doing the right thing.
Before this, the original implementation would loop forever if 'last' equal to INT_MAX.

astitcher commented on proton-c/src/core/transport.c:
```diff
@@ -1619,6 +1619,17 @@ static int pn_scan_error(pn_data_t *data, pn_condition_t *condition, const char
+  This operator, copied from code for the qpid cpp broker, gives the correct
+  result when comparing sequence numbers implemented in a signed integer type.
+*/
+static bool sequence_less_than ( pn_sequence_t a, pn_sequence_t b )
```
Why not return less than 0 for less than 0 for equal and greater than 0 for greater than (like strcmp)

astitcher replied:
ie code would be

static int sequence_cmp(pn_sequence_t a, pn_sequence_t b) {
    return a-b;
}

astitcher commented on proton-c/src/core/transport.c:
```diff
@@ -1648,7 +1659,7 @@ int pn_do_disposition(pn_transport_t *transport, uint8_t frame_type, uint16_t ch
                       pn_data_get_list(transport->disp_data) > 0);
 
-  for (pn_sequence_t id = first; id <= last; id++) {
+  for (pn_sequence_t id = first; sequence_less_than(id, last) || (id == last); id++) {
```
Then this test would be sequence_cmp(id, last)<=0

astitcher replied:
Of course this test will still do the wrong thing if the wrap-around is big enough (ie if first > last at teh start of the loop)

mgoulish commented:
I have updated to use Andrew's cleaner sequence_cmp() idea, and added a new change: preventing sign-extension from pn_sequence_t to uintptr_t in the calls to pn_hash_put() and pn_hash_get().   This sign extension appears to be happening in one pathway but not the other, so that , when we wrap around to negative p_sequence_t values, pn_hash_get() and pn_hash_put() are getting different hash values.

I know how this sounds -- but with this change my test got through the 2^31 barrier for the first time ever.
It sounds like heresy -- but -- eppur si muove.

This is probably not the end of the story -- I had runaway memory consumption later, but the test got to 2.25 billion messages before I had to kill it.  Well past the 2^32 barrier, which == 2.147 billion.

mgoulish commented:
By the way -- to be able to check this much faster, rather than running a test that has to count all the way up to 2^31 messages, I have been changing one line in engine.c, from this:
pn_delivery_map_init(&ssn->state.outgoing, 0);
to this:
pn_delivery_map_init(&ssn->state.outgoing, 2145483648);
// NOTE -- that number is 2^31minus 2 million.

However, after this new fix passed that shortened test several times, I ran a full-sized test, starting at zero.  That was successful also.

astitcher commented on proton-c/src/core/transport.c:
```diff
@@ -1620,12 +1622,12 @@ static int pn_scan_error(pn_data_t *data, pn_condition_t *condition, const char
   result when comparing sequence numbers implemented in a signed integer type.
 */
-static bool sequence_less_than ( pn_sequence_t a, pn_sequence_t b )
+static bool sequence_cmp ( pn_sequence_t a, pn_sequence_t b )
```
This should be ```inline```

astitcher commented on proton-c/src/core/transport.c:
```diff
@@ -82,9 +82,11 @@ void pn_delivery_map_free(pn_delivery_map_t *db)
   pn_free(db->deliveries);
 }
 
+#define BOTTOM_32_BITS 0x00000000FFFFFFFF
```
Isn't this just UINT32_MAX?

astitcher commented:
Having looked a bit more it seems less clear that the bug is in pn_hash!

Essentially the lower 32 bit masking is used as a more determinate way to cast the int32_t of the pn_sequence_t to the uintptr_t (uint64_t) of the hash code I don't understand (yet) why this casting sometimes gives different results, but I *strongly suggest* creating a better named inline function to express the intent rather than doing the masking with no real explanation.

Viz:

    static inline uintptr_t pni_sequence_make_hash(pn_sequence_t i) {
        return i & 0xFFFFFFFFUL;
    }

astitcher commented:
Also I *strongly suggest* you test this code on both 32 and 64 bit machines to make sure that the casting works equally well when uintptr_t is 32 vs 64 bits.

This is independent of whatever solution you end up with.

mgoulish commented:
Closing this in favor of doing a new one, with latest improvement.  New PR will have just one checkin.
